### PR TITLE
Allow calculating the layout behind a pointer

### PR DESCRIPTION
**Pull request by CAD97**
There was some discussion around allowing this previously.

This does make the requirement for raw pointers to have valid metadata exposed as part of the std API (as a safety invariant, not validity invariant), though I think this is not strictly necessarily required as of current. cc @rust-lang/wg-unsafe-code-guidelines 

Naming is hard; I picked the best "obvious" name I could come up with.

If it's agreed that this is actually a desired API surface, I'll file a tracking issue and update the attributes.

### CODE REVIEW

File: src/libcore/mem/mod.rs
Comment (LukasKalbertodt) [inline src/libcore/mem/mod.rs]:
```suggestion
/// ```
#![feature(layout_for_ptr)]
```

File: src/libcore/mem/mod.rs
Comment (LukasKalbertodt) [inline src/libcore/mem/mod.rs]:
```suggestion
/// ```
#![feature(layout_for_ptr)]
```

**Review by LukasKalbertodt — COMMENTED** [review]
Could you explain how `Layout::for_ptr` would be used? Are people supposed to create a dummy/aliasing pointer and pass it to the function? Is that really something useful? (Or just link to that previous discussion, so I can read up on it there.)

**Comment by RalfJung** [comment]
> This does make the requirement for raw pointers to have valid metadata exposed as part of the std API (as a safety invariant, not validity invariant), though I think this is not strictly necessarily required as of current.

I don't think that is a good decision -- and certainly it *is* a decision that requires FCP at least.

Given the general description of raw pointers as being geared towards unsafe code, I think having the safety invariant require more than the validity invariant is going against the expectation of most people.

So I'd prefer if this function were `unsafe`, with a safety precondition saying that the metadata on the raw pointers is "valid" or "proper" or whatever adjective we use. *If* we end up with stronger invariants for raw pointers later down the road, we can always remove the `unsafe` qualifier, that should be backwards-compatible.

**Comment by CAD97** [comment]
IIRC, the current proposed validity requirement for fat pointers would be enough for these functions. I only note this as a safety rather than validity requirement because the safety of these APIs only impacts the safety requirement. I'm perfectly happy to mark these APIs as `unsafe` until that question is resolved, though. (I agree that this should probably go through an FCP, though.)

I know the availability of `size_of_val`/`align_of_val` for pointers rather than references has been discussed previously, though I can't find any of said discussion currently.

The specific case that prompted me to actually do the legwork to see if this were possible is `liballoc::rc::data_offset`.

```rust
unsafe fn data_offset<T: ?Sized>(ptr: *const T) -> isize {
    // Align the unsized value to the end of the `RcBox`.
    // Because it is ?Sized, it will always be the last field in memory.
    // Note: This is a detail of the current implementation of the compiler,
    // and is not a guaranteed language detail. Do not rely on it outside of std.
    data_offset_align(align_of_val(&*ptr))
}
```

In this case the pointer is known to always be valid, but being able to directly query the layout of a type behind a pointer would eliminate the need to manifest a reference in this function.

Getting the layout behind a pointer also makes typed dealloc easier. Just as a quick example,

```rust
unsafe fn dealloc_t<T: ?Sized>(a: &mut impl AllocRef, ptr: NonNull<T>) {
    a.dealloc(ptr.cast(), Layout::for_ptr(ptr.as_ptr()))
}
```

The other primary potential usage I see for getting layout from the raw pointer rather than a reference is if the pointer is being passed around and the user wants to get the layout back without manifesting a reference, thus requiring grabbing exclusive access (`&mut`) or ruling out shared mutability (temporarily) (`&`).

To be completely honest, I'd be fine if the decision is that this isn't needed/desired. If some code has created a pointer that has valid pointer metadata but isn't valid as a reference, they _probably_ already know the layout (from allocation or when it was a valid reference); in that case this would only be allowing to retrieve said layout information back from the type system rather than a formal parameter.

**Comment by CAD97** [comment]
I've not touched the safety of the intrinsics (yet) but I marked the exposed API as conservatively `unsafe`. I'll let FCP decide any further details.

**Comment by RalfJung** [comment]
> IIRC, the current proposed validity requirement for fat pointers would be enough for these functions.

One proposal for validity requirements for vtables is that this is the same as `NonNull<()>`. That would certainly not be enough for these functions.

I am perfectly on board with adding these methods for raw pointers, that has a clear use. I am just saying we should be cautious about making these methods *safe*.

**Comment by LukasKalbertodt** [comment]
I know way too little about unsafe stuff to be a useful assignee for this PR. 

r? @RalfJung

**Comment by nikomatsakis** [comment]
I agree with @RalfJung that this precise point -- if I'm understanding correctly -- has been a point of contention. I recall discussing it with @Manishearth, for example, who felt strongly that one should be able to construct `*mut dyn Foo` pointers with a dummy/NULL vtable, so long as you knew that the pointer wouldn't be used.

**Comment by RalfJung** [comment]
> one should be able to construct *mut dyn Foo pointers with a dummy/NULL vtable, so long as you knew that the pointer wouldn't be used.

(Just to be clear, right now NULL specifically is insta-UB because raw wide pointers use the vtable as niche for layout optimizations. But we could conceivably change that.)

**Review by RalfJung — APPROVED** [review]
From a UCG perspective, this LGTM.

However, this is a public (but unstable) libstd addition, so maybe @rust-lang/libs should be involved?
Also, the new libs feature needs a tracking issue.

**Comment by Centril** [comment]
From @RalfJung's comments it seems like the language level component of the PR was removed cause of the addition of `unsafe`, so removing T-Lang accordingly.

**Comment by SimonSapin** [comment]
`unsafe fn`s should have a Safety section in their doc-comment that explains what requirements are expected of callers.

**Comment by CAD97** [comment]
What exactly should the safety section say for these functions? The technical requirement is that the pointer is thin, or that the fat pointer metadata is "enough" to determine size/align. And I'd argue that a safety requirement of "it is safe to do this" is worse than not having one. Of course, on top of this, with current defined Rust, it's not possible to violate this requirement.

A "safe to deref" requirement is over-constraining, though it is still looser than making a real reference. "Has valid metadata" is too loose, as validity refers to the validity invariant, and the point of it being `unsafe` is that there are some valid pointers which are not safe to use with these functions.

I definitely think these functions would already be useful, but I don't know how to specify their safety requirements while it's not possible to break them.

**Comment by SimonSapin** [comment]
Regardless of whether "valid" is the right term for these docs, as far as I can tell the requirements are something like:

* If `T` is `Sized` or an `extern type`, this function is always safe and returns a well defined result
* If `T` is a `[_]` slice type, this function is always safe but may return an arbitrary result if the computation overflows
* If `T` is a `dyn Trait` trait object, the pointer must have been originally created by unsizing coercion from a pointer to a concrete type

(In 1.41.0 on x64, `mem::size_of_val(slice::from_raw_parts(&0_u32, (1 << 62) + 1))` return 4.)

> I agree with @RalfJung that this precise point -- if I'm understanding correctly -- has been a point of contention. I recall discussing it with @Manishearth, for example, who felt strongly that one should be able to construct `*mut dyn Foo` pointers with a dummy/NULL vtable, so long as you knew that the pointer wouldn't be used.

But until we add something like [`<*const T>::from_raw_parts(data_ptr, metadata)`](https://github.com/rust-lang/rfcs/pull/2580) there is no way to do this without something transmute-like that makes assumptions about undocumented details of the memory layout of fat pointers, is there?

**Comment by SimonSapin** [comment]
I suppose `MaybeUninit::zeroed().assume_init()` could be a way, if we definitely decide that any bit pattern is valid for `*const dyn Trait` as long as it’s not dereferenced. (I’m skeptical we *should* decide that, though.)

**Comment by RalfJung** [comment]
> If T is a [_] slice type, this function is always safe but may return an arbitrary result if the computation overflows

Does "arbitrary" include undef/poison? Depending on how we lower this to LLVM, I think we should say that overflow is UB.

Also, we have to account for unsized structs with a `dyn Trait` tail (and similar). It's not the type `T` that is most relevant, but its "unsized tail".

> "Has valid metadata" is too loose, as validity refers to the validity invariant, and the point of it being unsafe is that there are some valid pointers which are not safe to use with these functions.

Hm yeah, that's why some people dislike the term "validity invariant". In some less formal docs we call it "properly initialized" or "initialization invariant", but I don't feel that solves this concern. I don't think we have a better term.

**Comment by SimonSapin** [comment]
My “something like” was a starting point. I think it’s fine to declare overflow UB in slice size computation. And yes, good point, a list of cases like this would need to account for DST structs.

**Comment by Manishearth** [comment]
>  I recall discussing it with @Manishearth, for example, who felt strongly that one should be able to construct `*mut dyn Foo` pointers with a dummy/NULL vtable, so long as you knew that the pointer wouldn't be used.

The main point was: unsafe code crops up in some weird places, and not having the ability to do this at all means that you can't mess with these in a step-by-step way. It's really hard to say what will and won't be necessary here. I feel like it would be easier to make it UB to _call_ any method with an invalid vtable on such a pointer: we already have this requirement for the pointer value itself.

MaybeUninit has _helped_ obviate a lot of cases like this, but I'm not sure if it's all of them.

**Comment by CAD97** [comment]
(writing this down before I forget) I

I think a good starting requirement for these functions is that the pointer describes a valid allocation/place (or a subset of one).

This is still useful for generic code. Just as a simple example, you could implement a `fn dealloc<T: ?Sized>(ptr: *mut T) { mem::dealloc(ptr, Layout::of_ptr(ptr)) }`. I don't recall the exact use case that tipped me over into making this PR, but I do know it was working with pointers to real allocations (just ones not known to be safe to turn into references).

This is an understandable conservative bound for what pointers are safe for this operation. When we have vocabulary to talk about pointer metadata and it's possible to make pointers that aren't safe to use in these functions, we can refine this bound.

-----

Adjacent note: the addition of this API surface is an opportunity to treat `extern type` in a more principled way. I don't know the exact status of `extern type` and `size`/`align_of`, other than the fact that `extern type` is treated as having unknown size/align and said functions purport to always provide a result. (Last I heard they'd panic? Is it different for the `of_val` functions?)

If we wanted to, we could return an `Option` from these functions instead.

**Comment by RalfJung** [comment]
@CAD97 so what about adapting some variant of @SimonSapin 's proposal (with adjustments as discussed) for the safety comment, so that we have something concrete to discuss?

**Comment by RalfJung** [comment]
Also @jonas-schievink you added needs-fcp; now that this just adds an unstable method does that still apply?

**Comment by CAD97** [comment]
> If `T` is a `[_]` slice type, this function is always safe but may return an arbitrary result if the computation overflows.

This might be too permissive if we end up allowing uninitialized `usize`. Then it'd be possible to do `let size = mem::uninit(); let ptr = slice_from_raw_parts(null(), size);` and get a slice pointer with uninitialized size.

> so what about adapting some variant of @SimonSapin 's proposal (with adjustments as discussed) for the safety comment, so that we have something concrete to discuss?

Whoops, I honestly thought I had added a safety section already. Because of the above worry, I've added the safety statement as the conservative

> This function is only safe to call with a pointer to a valid allocation.
> In other words, if it would be safe to dereference the pointer as a
> [`*const MaybeUninit<T>`][MaybeUninit], then this function is safe.
>
> This restriction may be loosened in the future.

It's not a perfect definition (`MaybeUninit` is for sized types only, you can argue `slice_from_raw_parts(&[], 1_000)` is a pointer to a valid allocation (but with the wrong pointer metadata), and it's overly conservative), but it gives a starting place.

[MaybeUninit]: <https://doc.rust-lang.org/core/mem/union.MaybeUninit.html>

> now that this just adds an unstable method does [needs-fcp] still apply?

s/unstable/unsafe; this was always marked as unstable. this is now explicitly avoiding implying anything new about the language and just exposing new unsafe functions. You could still argue it should FCP, though, because it does touch intrinsics (which has an impact on the spec to support the stdlib for alternative implementations) and adds un-RFC'd API surface (though still unstable).

**Comment by RalfJung** [comment]
> This might be too permissive if we end up allowing uninitialized usize. 

Yeah, see the later comment that says we should even make overflows UB.
For now, I think we should require this to be an initialized integer that does not overflow when doing the size computation.

File: src/libcore/alloc.rs
Comment (RalfJung) [inline src/libcore/alloc.rs]: Wait, you want to require the pointer to be dereferencable? That's much stronger than anything we discussed, I feel, and also what is even the point of this method?

Comment (CAD97) [reply src/libcore/alloc.rs]: Note that this is only the current, most conservative starting place. [I did mention the intent here](https://github.com/rust-lang/rust/pull/69079#issuecomment-593205078). Note that the use of `MaybeUninit` here is intending to make it such that the pointee does not need to be initialized, just exist.

Comment (RalfJung) [reply src/libcore/alloc.rs]: This is quite confusing though as `MaybeUninit<T>` does not exist unless `T` is sized.

Comment (CAD97) [reply src/libcore/alloc.rs]: Alright, I've got a looser definition closer to @SimonSapin's that should be clearer now.

**Comment by CAD97** [comment]
To keep top level discussion accurate, safety note is now currently:

> This function is only safe to call if one of the following conditions hold:
>
> - The pointer is safe to dereference,
> - `T` is `Sized`, or
> - the unsized tail of `T` is a [slice], and the place described by the pointer would fit in `usize` memory, though the described place does not need to actually be allocated.
>
> This restriction may be loosened in the future.

**Comment by SimonSapin** [comment]
Nit: does “fit in `usize` memory” mean the size computation does not overflow `usize::MAX` bytes? I assume so based on previous comments in this thread, but by itself this isn’t clear to me.

**Comment by RalfJung** [comment]
Given that the only alternative unsized tail is a trait object, maybe it is easier to just spell out that trait objects need to have a valid vtable matching the dynamic type of the data they point to?

Well and I guess there are extern types but those should be easy. I think a case distinction by unsized tail is more comprehensible than your current approach:

* if there is no unsized tail, this function is always safe
* if the unsized tail is a slice, ...
* if the unsized tail is a trait object, ...
* if the unsized tail is an extern type, ...
* there are no other cases

**Comment by RalfJung** [comment]
I also agree with @SimonSapin, it should state explicitly that doing size computation with `usize` must not overflow. Which reminds me, are we sue it will work for `usize`, or do we have to require that it does not overflow in `isize`?

Also, there is another possible overflow: with a type like `(i32, [i32])`, the slice itself could be fine in terms of length, but then adding the 4 bytes of the sized prefix could overflow.

**Comment by SimonSapin** [comment]
Re `(i32, [i32])` yes, “the size computation” should be phrased to refer to the whole computation `size_of_ptr` does, not just the unsized tail.

Speaking of, the name `size_of_ptr` sounds to me like it would always return 8 or 16 bytes (on 64-bit platforms). This computes the size of the pointed value, not of the pointer itself.

**Comment by CAD97** [comment]
> Re `(i32, [i32])` yes, “the size computation” should be phrased to refer to the whole computation size_of_ptr does, not just the unsized tail.

That's the intent of the current wording, "the place described by the pointer would fit in `usize` memory," and in fact as stated it's more restrictive than just that. Specifically, it puts its restriction on the pointer's described place (i.e. the whole pointee), not just the unsized tail, and because of the "fits in memory" requirement, `(usize::MAX, 2) as *mut [u8]` would be invalid to call with. (I don't know if that case is or isn't actually UB with the intrinsic as currently implemented, but it is the more conservative option that allows it to be implemented as just `one-past-the-end - base`.)

> Speaking of, the name `size_of_ptr` sounds to me like it would always return 8 or 16 bytes (on 64-bit platforms). This computes the size of the pointed value, not of the pointer itself.

What about `Layout::for_ptr`? Other options I considered: `[for|of]_pointee`, `of_raw_val`/`of_val_raw`, `of_raw`.

I think I agree in retrospect that `size_of_ptr` should (if it exists) be `fn size_of_ptr<T>() => size_of::<*const T>()`.

> I think a case distinction by unsized tail is more comprehensible than your current approach

Because `extern type` is still unstable and the language doesn't yet have language to talk about trait object pointers' metadata, I'd prefer to avoid talking directly about them. After all, the _only_ way to construct them as of current is by coersion from a reference.

I definitely think "pointer was coerced from a reference" is the simplest baseline that a pointer is valid to do these operations on. Though I do suppose that it does accidentally exclude pointers acquired from `Box::into_raw` and other "smart pointer" types. (Note that even if you have a pointer that is safe to dereference where the pointee is fully safely initialized, that [does not mean you necessarily have the ability to create a reference](https://github.com/CAD97/pointer-utils/pull/34).)

This vague "safely describes a real place that existed at some point" heuristic is the biggest one, and I think the most useful one. (Enough so that I think the functions would be useful even if they stopped there.) If that is specified as safe, then the only remaining cases remaining to support (that are possible in today's Rust) are dangling pointers to sized types and dangling pointers to slice types.

If we want to define it on a case over the unsized tail, how about something like

- If `T` is sized, this function is safe.
- Otherwise, if the unsized tail of `T` is a slice, [the pointer's described place fits in memory].
- Otherwise, the pointer must be derived from a reference [or smart pointer type].

Though the exact phrasing of the bracketed bits is up to clarification.

**Comment by RalfJung** [comment]
> That's the intent of the current wording, "the place described by the pointer would fit in usize memory," and in fact as stated it's more restrictive than just that. Specifically, it puts its restriction on the pointer's described place (i.e. the whole pointee), not just the unsized tail, and because of the "fits in memory" requirement, (usize::MAX, 2) as *mut [u8] would be invalid to call with. (I don't know if that case is or isn't actually UB with the intrinsic as currently implemented, but it is the more conservative option that allows it to be implemented as just one-past-the-end - base.)

I think it would be really odd to make any requirement for the `data` part of the pointer here.

One could argue that this helps for potential future custom DST proposals, but those are so far out and so unclear, I think it would be a mistake to try and design for them now.

> `of_raw_val`/`of_val_raw`

These two are my favorite bikeshed colors.

> Otherwise, the pointer must be derived from a reference [or smart pointer type].

Given what I said above about the data ptr not mattering, I think I'd prefer something that said that the vtable is one that was created by an actual unsizing coercion ("a valid vtable"), and moreover computing the size of the full thing (trait object tail plus prefix) does not overflow `usize`.

We should carefully check codegen to see if we don't have to instead say "does not overflow `isize`". Or to be conservative we could just start with that.

**Comment by CAD97** [comment]
Alright, I updated the safety docs again as well as the names. New safety docs:

> This function is only safe to call if one of the following conditions hold:
>
> - If `T` is `Sized`, this function is always safe to call.
> - If the unsized tail of `T` is:
>     - a [slice], then the length of the slice tail must be an intialized
>       integer, and the size of the described `T` must fit in `isize`.
>     - a [trait object], then the vtable part of the pointer must point
>       to a valid vtable acquired by an unsizing coersion.
>     - otherwise, it is conservatively not allowed to call this function.
>
> This restriction may be loosened in the future.

File: src/libcore/alloc.rs
Comment (RalfJung) [inline src/libcore/alloc.rs]: I don't entirely see the point of emphasizing this, isn't that basically implicit anywhere?

File: src/libcore/alloc.rs
Comment (RalfJung) [inline src/libcore/alloc.rs]:
```suggestion
    ///       integer, and the size of the *entire value* must fit in `isize`.
```

File: src/libcore/alloc.rs
Comment (RalfJung) [inline src/libcore/alloc.rs]:
```suggestion
    ///       to a valid vtable acquired by an unsizing coersion, and the size
    ///       of the *entire value* (dynamic tail length + statically sized prefix)
    ///       must fit in `isize`.
```

File: src/libcore/mem/mod.rs
Comment (RalfJung) [inline src/libcore/mem/mod.rs]: This and the below markdown reference are dangling.

File: src/libcore/mem/mod.rs
Comment (RalfJung) [inline src/libcore/mem/mod.rs]: More dangling markdown references.

File: src/libcore/alloc.rs
Comment (RalfJung) [inline src/libcore/alloc.rs]: So the plan is to not allow this function on extern types? Or did you just not mention them because they are unstable?

If it is the latter, I feel we should mention them somehow -- maybe just saying that this is talking about an unstable language feature. But the question is inevitably going to come up.

Comment (CAD97) [reply src/libcore/alloc.rs]: I believe it's "more proper" to disallow asking for extern types through this API, as they don't actually have a knowable size/align (from Rust's POV). However, depending on the final decision for calling the reference-based versions, the most consistent result may even change; perhaps we can use these functions to return an `Option` that's `None` for extern types.

IOW, I've not said that `extern type` is defined to give to this function because I honestly don't know what semantics are best to give them.

File: src/libcore/mem/mod.rs
Comment (CAD97) [reply src/libcore/mem/mod.rs]: These are actually not dangling; they're defined above, in the text copied from the non-raw API. Though perhaps I _should_ move those anchors to the bottom.

File: src/libcore/mem/mod.rs
Comment (CAD97) [reply src/libcore/mem/mod.rs]: These are in fact dangling, and the ones on `Layout::of_value_raw` are broken (for some reason; apparently I don't actually know how to get to the right place from there).

File: src/libcore/alloc.rs
Comment (RalfJung) [reply src/libcore/alloc.rs]: It would be weird to have a semantics that's any different from `size_of_val`, wouldn't it? `extern type` has no metadata, just like sized types, after all.

Comment (CAD97) [reply src/libcore/alloc.rs]: Of course, `size_of_val`'s semantics are [to panic in the face of `extern type` (most likely)](https://github.com/rust-lang/rust/issues/49708)

Comment (RalfJung) [reply src/libcore/alloc.rs]: Seems reasonable then to document that this will behave just like `size_of_val`?

Comment (CAD97) [reply src/libcore/alloc.rs]: I have "but may panic or otherwise return the wrong value, as the extern type's layout is not known" currently. Given extern type is still unstable and the implementation of `[size|align]_of_val` don't yet match (what I _think_ is) consensus, I think this loose "safe but otherwise unspecified" is _probably_ enough to land these functions, and the wording can be re-addressed by the PR clarifying the actual intended behavior.

Comment (RalfJung) [reply src/libcore/alloc.rs]: I agree having such a vague wording makes sense. But people will also wonder "is there any difference to calling `size_of_val`", won't they?

Comment (CAD97) [reply src/libcore/alloc.rs]: I added "This is the same behavior as [`fn`] on a reference to an extern type tail", so this should be adequate now.

File: src/libcore/alloc.rs
Comment (RalfJung) [inline src/libcore/alloc.rs]:
```suggestion
    /// This function is only safe to call if the following conditions hold:
```

File: src/libcore/mem/mod.rs
Comment (RalfJung) [inline src/libcore/mem/mod.rs]:
```suggestion
/// This function is only safe to call if the following conditions hold:
```

File: src/libcore/mem/mod.rs
Comment (RalfJung) [inline src/libcore/mem/mod.rs]:
```suggestion
/// This function is only safe to call if the following conditions hold:
```

**Review by RalfJung — APPROVED** [review]
LGTM (modulo the last nits), thanks for enduring!

@SimonSapin what do you think?

**Comment by RalfJung** [comment]
@CAD97 okay @Mark-Simulacrum says this is okay for t-libs. Could you squash the commits?

**Comment by CAD97** [comment]
Squashed.

**Comment by RalfJung** [comment]
Thanks! @bors r+

**Comment by arslan-raza-143** [comment]
> There was some discussion around allowing this previously.
> 
> 
> 
> This does make the requirement for raw pointers to have valid metadata exposed as part of the std API (as a safety invariant, not validity invariant), though I think this is not strictly necessarily required as of current. cc @rust-lang/wg-unsafe-code-guidelines 
> 
> 
> 
> Naming is hard; I picked the best "obvious" name I could come up with.
> 
> 
> 
> If it's agreed that this is actually a desired API surface, I'll file a tracking issue and update the attributes.

**Review by arslan-raza-143 — COMMENTED** [review]
@CAD97

**Pull request merged**
